packages/cozy-stack-client/src/FileCollection.js

```diff
@@ -253,6 +253,37 @@ class FileCollection extends DocumentCollection {
     return resp.links.related
   }
 
+  /**
+   * Checks if the file belongs to the parent's hierarchy.
+   *
+   * @param  {string}  childFileID  The file ID of the child
+   * @param  {string}  childDirID   The dirID of the child
+   * @param  {string}  childPath    The child path
+   * @param  {string}  parentID     The parent's document id
+   * @return {boolean}              Whether the file is a parent's child
+   */
+  async isChild(childFileID, childDirID, childPath, parentID) {
+    if (childFileID === parentID || childDirID === parentID) {
+      return true
+    }
+    const paths = childPath.split('/')
+    paths.shift() // The path starts with '/'
+
+    const targetsPath = ['/' + paths[0]]
+    for (let i = 1; i < paths.length; i++) {
+      const newPath = targetsPath[i - 1] + '/' + paths[i]
+      targetsPath.push(newPath)
+    }
+
+    // Look for all hierarchy in parallel
+    const results = await Promise.all(
```

Comment:
This will wait for the last request to fullfill even if we have a successfull one already fullfilled. Could we implement a shortcut there ?

Reply:
How do you see the shortcut? We didn't see any with @CPatchane as we cannot cancel promises

Reply:
> The Promise.race() method returns a promise that fulfills or rejects as soon as one of the promises in an iterable fulfills or rejects, with the value or reason from that promise.

something like ? 
```
await Promise.race(targetsPath.map(path => this.statByPath(path).then(result => if(result.data._id === parentID) resolve()))
```

Reply:
Something like (not tested)

```js
return new Promise( async (resolve) => {
  await Promise.all(
    targetsPath.map(
      async (path) => {
        const stat = await this.statByPath(path)
        if (stat.data._id == parentID) {
          resolve(true)
        }
      }
    )
  )
  resolve(false)
})
```

Reply:
Promise.race takes the first to resolve or reject. We want to take the first to resolve and discard all rejections, or take the first to resolve true and discard those which resolve to false.

We probably need the Promise.all, but we can return without waiting for it to finish

Reply:
@edas Nice, your solution seems to work!

Reply:
I like the `race` idea and I think it would be cleaner if the race part was factored out of the business logic.


```js
const raceWithCondition = (promises, predicate, default) => {
  return new Promise((resolve, reject) => {
    let rejections = 0
    const handleError = err => ++rejections === promises.length && reject(e)
    for (const p of promises) {
      p.then(res => predicate(res) && resolve(res)).catch(handleError)
    }
    Promise.all(promises).then(() => resolve(default))
  })
}

const res = await raceWithCondition(
  targetsPath.map(path => this.statByPath(path)),
  stat => stat.data._id == parentID
)
```

Reply:
Love it, thanks

Reply:
I was trying to see if bluebird had this natively. I stumbled upon this little comment in http://bluebirdjs.com/docs/api/promise.race.html which is interesting :

> This method is only implemented because it's in the ES6 standard. If you want to race promises to fulfillment the .any method is more appropriate as it doesn't qualify a rejected promise as the winner. It also has less surprises: .race must become infinitely pending if an empty array is passed but passing an empty array to .any is more usefully a RangeError